File: util-rs/src/scc.rs
#[derive(Clone)]
struct Edge {
    from: usize,
    to: usize,
}

struct Graph {
    out_edges: Vec<Vec<Edge>>,
    in_edges: Vec<Vec<Edge>>,
}

impl Graph {
    #[allow(dead_code)]
    fn new(n: usize) -> Graph {
        Graph {
            out_edges: vec![vec![]; n],
            in_edges: vec![vec![]; n],
        }
    }

    #[allow(dead_code)]
    fn add_edge(&mut self, from: usize, to: usize) {
        self.out_edges[from].push(Edge { from, to });
        self.in_edges[to].push(Edge { from, to });
    }

    #[allow(dead_code)]
    fn size(&self) -> usize {
        self.out_edges.len()
    }

    #[allow(dead_code)]
    fn get_out_edges(&self, from: usize) -> &Vec<Edge> {
        &self.out_edges[from]
    }

    #[allow(dead_code)]
    fn get_in_edges(&self, to: usize) -> &Vec<Edge> {
        &self.in_edges[to]
    }
}

#[allow(dead_code)]
fn dfs(g: &Graph, v: usize, visited: &mut Vec<bool>, vs: &mut Vec<usize>) {
    visited[v] = true;

    for edge in g.get_out_edges(v) {
        if !visited[edge.to] {
            dfs(g, edge.to, visited, vs);
        }
    }

    vs.push(v);
}

#[allow(dead_code)]
fn rev_dfs(g: &Graph, v: usize, visited: &mut Vec<bool>, vs: &mut Vec<usize>) {
    visited[v] = true;
    vs.push(v);

    for edge in g.get_in_edges(v) {
        if !visited[edge.from] {
            rev_dfs(g, edge.from, visited, vs);
        }
    }
}

// 強連結成分分解
#[allow(dead_code)]
fn scc(g: &Graph) -> Vec<Vec<usize>> {
    let mut vs = vec![];
    {
        let mut visited = vec![false; g.size()];
        for v in 0..g.size() {
            if !visited[v] {
                dfs(g, v, &mut visited, &mut vs);
            }
        }
    }

    let mut ret = vec![];
    {
        let mut visited = vec![false; g.size()];
        for &v in vs.iter().rev() {
            if !visited[v] {
                let mut component = vec![];
                rev_dfs(g, v, &mut visited, &mut component);
                ret.push(component);
            }
        }
    }

    ret
}

// 2-sat
#[allow(dead_code)]
struct TwoSat {
    g: Graph,
}

impl TwoSat {
    #[allow(dead_code)]
    fn new(n: usize) -> TwoSat {
        TwoSat {
            g: Graph::new(2 * n),
        }
    }

    // size()より小さいとfalse, size()以上だとtrue
    #[allow(dead_code)]
    fn to_v(&self, a: usize, f: bool) -> usize {
        if f {
            a + self.size()
        } else {
            a
        }
    }

    #[allow(dead_code)]
    fn add(&mut self, a: usize, fa: bool, b: usize, fb: bool) {
        self.g.add_edge(self.to_v(a, !fa), self.to_v(b, fb));
        self.g.add_edge(self.to_v(b, !fb), self.to_v(a, fa));
    }

    #[allow(dead_code)]
    fn size(&self) -> usize {
        self.g.size() / 2
    }

    #[allow(dead_code)]
    fn solve(&self) -> Option<Vec<bool>> {
        let components = scc(&self.g);

        let mut ret = vec![false; self.size()];
        let mut idx = vec![components.len(); self.size()];
        for i in 0..components.len() {
            for &v in &components[i] {
                let t = v % self.size();

                if idx[t] == i {
                    // negation is already appeared in same component.
                    return None;
                }

                if idx[t] == components.len() {
                    idx[t] = i;
                    // vの否定を立てる
                    ret[t] = v < self.size();
                }
            }
        }

        Some(ret)
    }
}

mod test {
    #[allow(unused_imports)]
    use super::*;

    #[test]
    fn test_scc() {
        let mut g = Graph::new(6);
        g.add_edge(0, 1);
        g.add_edge(1, 0);
        g.add_edge(1, 2);
        g.add_edge(2, 3);
        g.add_edge(3, 4);
        g.add_edge(4, 5);
        g.add_edge(5, 3);

        let mut components = scc(&g);
        assert!(components.len() == 3);

        for component in components.iter_mut() {
            component.sort();
        }
        assert!(components[0] == vec![0, 1]);
        assert!(components[1] == vec![2]);
        assert!(components[2] == vec![3, 4, 5]);
    }

    #[test]
    fn test_two_sat() {
        let mut sat = TwoSat::new(3);
        sat.add(0, true, 1, true);
        sat.add(1, true, 2, true);
        sat.add(0, true, 2, false);

        assert!(sat.solve() == Some(vec![false, true, false]))
    }
}

File: cpp/util-cpp/scc.cpp
#include <bits/stdc++.h>

#include <boost/range/adaptors.hpp>
#include <boost/range/irange.hpp>

using namespace std;
using namespace boost;
using namespace boost::adaptors;

struct Edge {
    int64_t from;
    int64_t to;
};

class Graph {
   public:
    explicit Graph(int64_t n) : out_edges_(n), in_edges_(n) {}

    void add_edge(int64_t from, int64_t to) {
        out_edges_[from].push_back(Edge{from, to});
        in_edges_[to].push_back(Edge{from, to});
    }

    int64_t size() const { return out_edges_.size(); }

    const vector<Edge>& out_edges(int64_t from) const {
        return out_edges_[from];
    }

    const vector<Edge>& in_edges(int64_t to) const { return in_edges_[to]; }

   private:
    vector<vector<Edge>> out_edges_;
    vector<vector<Edge>> in_edges_;
};

void dfs(const Graph& g, int64_t v, vector<bool>& visited,
         vector<int64_t>& vs) {
    visited[v] = true;

    for (const auto& edge : g.out_edges(v)) {
        if (!visited[edge.to]) {
            dfs(g, edge.to, visited, vs);
        }
    }

    vs.push_back(v);
}

void rev_dfs(const Graph& g, int64_t v, vector<bool>& visited,
             vector<int64_t>& vs) {
    visited[v] = true;
    vs.push_back(v);

    for (const auto& edge : g.in_edges(v)) {
        if (!visited[edge.from]) {
            rev_dfs(g, edge.from, visited, vs);
        }
    }
}

// 強連結成分分解
vector<vector<int64_t>> scc(const Graph& g) {
    vector<bool> visited(g.size());

    vector<int64_t> vs;
    for (auto v : irange(0L, g.size())) {
        if (!visited[v]) {
            dfs(g, v, visited, vs);
        }
    }

    visited.assign(g.size(), false);
    vector<vector<int64_t>> ret;
    for (auto v : vs | reversed) {
        if (!visited[v]) {
            ret.emplace_back();
            rev_dfs(g, v, visited, ret.back());
        }
    }

    return ret;
}

// 2-sat
class TwoSat {
   public:
    explicit TwoSat(int64_t n) : g(2 * n) {}

    void add(int64_t a, bool fa, int64_t b, bool fb) {
        g.add_edge(to_v(a, !fa), to_v(b, fb));
        g.add_edge(to_v(b, !fb), to_v(a, fa));
    }

    int64_t size() const { return g.size() / 2; }

    vector<bool> solve() const {
        const auto& components = scc(g);

        vector ret(size(), false);
        vector idx(size(), components.size());
        for (auto i : irange(0uL, components.size())) {
            for (auto v : components[i]) {
                auto t = v % size();

                if (idx[t] == i) {
                    return {};
                }

                if (idx[t] == components.size()) {
                    idx[t] = i;
                    // vの否定を立てる
                    ret[t] = (v < size());
                }
            }
        }

        return ret;
    }

   private:
    // size()より小さいとfalse, size()以上だとtrue
    int64_t to_v(int64_t a, bool f) const { return a + f * size(); }

    Graph g;
};